Fix - Workspace - Currencies do not load after transition from OldDot when other user logged in ND #31789

AmjedNazzal opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details

We are adding a check for when the user uses short lived token while deeplinking from OD, this is to solve the issue where currencies don't load in these cases.

As per the discussion in the PR, we have come to the conclusion that this bug is not reproducible on Android/iOS native in any typical workflow. This is evident in the component App.ts, where the beginDeepLinkRedirect function explicitly calls Browser.openRouteInDesktopApp. This suggests that there is no natural workflow where deep link redirection could result in the URL opening in the Android/iOS native app.

### Fixed Issues

$ https://github.com/Expensify/App/issues/31789
PROPOSAL: https://github.com/Expensify/App/issues/31789#issuecomment-1825429017


### Tests

For web/mweb:

1) Login to OD from with account A
2) Login to ND from account B
3) From OD go to Settings > Workspaces and click on a workspace
4) Wait until the browser address bar gets populated with the url and stop the loading 
5) Replace new.expensify with dev.new.expensify.com:8082 in the address bar url
5) Once the app loads and the workspace settings are mounted, check that currencies have loaded

For MacOS Desktop:

1) Follow the same steps as web for steps 1-3
2) Once ND loads, it will prompt you to open new expensify in your local Electron build
3) Accept the prompt and wait until the workspace settings load in your desktop app

- [x] Verify that no errors appear in the JS console

### Offline tests

No offline test

### QA Steps

1) Login to OldDot (staging) from with account A
2) Login to NewDot (staging) from account B
3) From OldDot, go to Settings > Workspaces and click on a workspace
4) Once the app loads and the workspace settings are mounted, check that currencies have loaded

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

N/A

</details>

<details>
<summary>Android: mWeb Chrome</summary>


https://github.com/Expensify/App/assets/74202751/ae123fd4-42eb-46df-b805-601854079ef7


</details>

<details>
<summary>iOS: Native</summary>

N/A

</details>

<details>
<summary>iOS: mWeb Safari</summary>


https://github.com/Expensify/App/assets/74202751/811dd3b2-6a28-481d-9071-4c56953cf279


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/Expensify/App/assets/74202751/fa40c24a-a5c8-4674-877a-baed2fab4a8f


</details>

<details>
<summary>MacOS: Desktop</summary>


https://github.com/Expensify/App/assets/74202751/ae288876-d54f-464e-8875-b98b823fddd8


</details>

AmjedNazzal committed:
Fix for issue #31789

AmjedNazzal commented:
Note: I have considered the use of union type as suggested but upon looking into it, changing authToken into a union type will mean that there will be changes needed accross multiple areas of the app because the use of authToken would change and typescript will throw errors otherwise. So this made me wonder, is it worth making changes in the current code and future code whenever authToken is used just for the sake of this somewhat edgecase localized issue we are trying to solve? let me know your thoughts on this.

cubuspl42 commented:
> Note: I have considered the use of union type as suggested but upon looking into it, changing authToken into a union type will mean that there will be changes needed accross multiple areas of the app because the use of authToken would change and typescript will throw errors otherwise. So this made me wonder, is it worth making changes in the current code and future code whenever authToken is used just for the sake of this somewhat edgecase localized issue we are trying to solve? let me know your thoughts on this.

I don't think this union was a good idea anymore. You're right, let's keep this as is.

cubuspl42 commented on src/libs/actions/Session/index.ts:
```diff
@@ -331,6 +331,15 @@ function signInWithShortLivedAuthToken(email: string, authToken: string) {
                 isLoading: true,
             },
         },
+        // We are making a temporary modification to 'signedInWithShortLivedAuthToken' to ensure that 'App.openApp' will be called at least once
```
Are we fine with `App.openApp` being called twice or more?

AmjedNazzal replied:
ideally we don't want multiple API calls but in this instance we aren't adding anything that in itself will cause App.openApp to be called multiple times, but if that happens as a side effect from something else it shouldn't be a problem since we are just doing onyx merge

cubuspl42 commented on src/libs/actions/Session/index.ts:
```diff
@@ -331,6 +331,15 @@ function signInWithShortLivedAuthToken(email: string, authToken: string) {
             },
         },
+        // We are making a temporary modification to 'signedInWithShortLivedAuthToken' to ensure that 'App.openApp' will be called at least once
+        // Subsequently, we revert it back to the default value (undefined) in 'successData' to ensure the user is logged out on refresh
```
I think that this comment could be moved to the place where this happens

AmjedNazzal replied:
Sure I will move that

AmjedNazzal committed:
Update index.ts

cubuspl42 commented on src/libs/actions/Session/index.ts:
```diff
@@ -331,8 +331,17 @@ function signInWithShortLivedAuthToken(email: string, authToken: string) {
     ];
 
+    // Subsequently, we revert it back to the default value of signedInWithShortLivedAuthToken (undefined) in 'successData' to ensure the user is logged out on refresh
     const successData: OnyxUpdate[] = [
```
Currently, `successData` and `failureData` is the same object. Could we store it in a single variable, like `completedData`?

I don't think we have a conventional name for such a thing, "completed" is the best thing that came to my mind. Could also be `resultData`?

AmjedNazzal replied:
We can do something like this: 

```
const commonData: OnyxUpdate[] = [
        {
            onyxMethod: Onyx.METHOD.MERGE,
            key: ONYXKEYS.ACCOUNT,
            value: {
                isLoading: false,
            },
        },
        {
            onyxMethod: Onyx.METHOD.MERGE,
            key: ONYXKEYS.SESSION,
            value: {
                signedInWithShortLivedAuthToken: undefined,
            },
        },
    ];
    
    const successData = [...commonData];
    const failureData = [...commonData];
```
    
        
What do you think?

AmjedNazzal replied:
We can do something like this: 

```
const commonData: OnyxUpdate[] = [
        {
            onyxMethod: Onyx.METHOD.MERGE,
            key: ONYXKEYS.ACCOUNT,
            value: {
                isLoading: false,
            },
        },
        {
            onyxMethod: Onyx.METHOD.MERGE,
            key: ONYXKEYS.SESSION,
            value: {
                signedInWithShortLivedAuthToken: undefined,
            },
        },
    ];
    
    const successData = [...commonData];
    const failureData = [...commonData];
```
    
        
What do you think?

AmjedNazzal replied:
Sure, we can use completed instead of common since whether or not it was success or failure it's still completed so it would make sense

AmjedNazzal replied:
I guess my only wonder here is, would we never add something different for successData within signInWithShortLivedAuthToken?

cubuspl42 replied:
`doneData` ? `finalData` ?

cubuspl42 replied:
It's literal copy-paste of a dozen of lines. I think it's obvious that they should be merged. We could even pass `failureData: successData` with a comment like `// Failure data and success data are same in this case`.

I'm fine with `resolutionData`, I'm also fine with `failureData: successData`, but please de-duplicate this. If success data and failure data start to differ in the future, it can by trivially changed.

AmjedNazzal replied:
@cubuspl42 Are you suggesting to pass `resolutionData` alongside `optimisticData` without defining `successData` and `failureData` seperately? I'm asking because I'm not very familure with how expensify is handling success and failure scenarious after the API call is made

AmjedNazzal replied:
Let me know what you think of the current changes

cubuspl42 replied:
Looks good 👍

AmjedNazzal committed:
Fix - issue #31789 adjusting comments

AmjedNazzal committed:
Fix - issue #31789 merging success and failure data

cubuspl42 commented:
First, use OldDot and NewDot, don't confuse new people in the project too much (these terms are already cryptic for newcomers)

> Wait until the browser address bar gets populated with the url and stop the loading then replace new.expensify with staging.new.expensify

It doesn't seem to be necessary: [(video)](https://private-user-images.githubusercontent.com/78819774/285280650-25646ed7-6d70-4d78-a9f3-420f5ace77e9.mp4?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE0Mzg0NzYsIm5iZiI6MTcwMTQzODE3NiwicGF0aCI6Ii83ODgxOTc3NC8yODUyODA2NTAtMjU2NDZlZDctNmQ3MC00ZDc4LWE5ZjMtNDIwZjVhY2U3N2U5Lm1wND9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjAxVDEzNDI1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQwYTFmMDJlMGI2OTY2MjAzNWRlMTc5OTE1MTkxNmI1OTU3ZTFjNDkxMTM3OTg2ZWVhZjQ3YmI1NjQ5NTNjOTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2e6y_gu13dwdIFlJvHheyvlxl2yQl3uikGv4TCIlT5M)

Replace "OD" with "OldDot (staging)" in the beginning and we should be good

AmjedNazzal committed:
Merge branch 'Expensify:main' into Issue31789

AmjedNazzal committed:
Fix - Issue #31789 - Lint fix

cubuspl42 commented:
In "Tests", would you specify what exact steps did you follow on non-Web platforms? Including any tricks and shell commands.

AmjedNazzal commented:
@cubuspl42 I've added steps for other platforms, let me know how it looks

cubuspl42 commented:
I have this issue:

https://github.com/Expensify/App/assets/2590174/bdcdbab1-1c8b-4bd7-a85f-bdc84b686f52

AmjedNazzal commented:
@cubuspl42 from the video it seem like you are getting an API error trying to open the workspace from OD before you even try using it on the native app correct? are you sure it's related to our PR here? because in the video it seem like the issue is happening on your end between OD and ND production?

AmjedNazzal committed:
Merge branch 'Expensify:main' into Issue31789

AmjedNazzal commented:
Okay I managed to reproduce the issue you are having on my physical android device, but I have noticed the issue happens with or without the changes we made in this PR so it must be an Android issue that was always there, either way I will look into the cause and let you know what I find.

AmjedNazzal commented:
@cubuspl42 This issue seem to be more about Android security features, I've been doing some tests and it seem like clicking a workspace in OD, whether it's from OD app or website, will not redirect to the native app in the normal natural flow because clicking it will always open ND in a browser so do we really need to worry about this? is there any case at all where clicking workspace from OD would open up the native app?

cubuspl42 commented:
I'm just trying to ensure we test the Android platform properly!

> is there any case at all where clicking workspace from OD would open up the native app?

I just investigated it. It _shouldn't_ open directly.

https://new.expensify.com/.well-known/assetlinks.json

https://github.com/Expensify/App/blob/fe6a2cb59dd992a2ef7b6b4f21130ca1503b5b5a/android/app/src/main/AndroidManifest.xml#L58-L74

cubuspl42 commented:
> from the video it seem like you are getting an API error trying to open the workspace from OD

It's not an API error, I think, I just aborted all requests as soon as possible, as I wanted to test opening the link directly in the native app.

AmjedNazzal commented:
@cubuspl42  Got you, what would you like to do about this? do I dive into this and investigate why Android is rejecting `signInWithShortLivedAuthToken`? or do you think since it wouldn't normally open in the native app we shouldn't worry about this?

cubuspl42 commented:
I'm just trying to ensure we properly test the Android flow, so no regressions arise. I don't have any greater ambitions here.

How did you figure out that the exact form of the `new-expensify:/` link? I know that it has the same form as the URL handled by the NewDot web app, but it wasn't guarnteed.

AmjedNazzal commented:
@cubuspl42 I actually didn't figure it out myself, that was something I found out in the slack channel where a c+ mentioned this link, if you want I can link you that discussion

cubuspl42 commented:
Sure, share it please

AmjedNazzal commented:
https://expensify.slack.com/archives/C01GTK53T8Q/p1698849973538749?thread_ts=1698848483.535709&cid=C01GTK53T8Q

AmjedNazzal committed:
Merge branch 'Expensify:main' into Issue31789

AmjedNazzal commented:
@cubuspl42 I actually did test and I could not get the deeplinking from OD to ever open ND, I tried production apps both OD and ND and OD app will always open up the browser, same with OD in the browser, it never attempts to launch the native app, so you are correct in saying that the android test is actually testing nothing because it doesn't seem like this issue we are working on would ever occur in native apps anyway unless we make changes that promts OD links to open in the native ND, which I don't know if it will ever happen since commonly apps will open links in the browser anyway.

AmjedNazzal commented:
@cubuspl42 any updates? should we remove the native tests and state the bug/fix is not applicable to native app? or should we ask on slack?

AmjedNazzal committed:
Merge branch 'Expensify:main' into Issue31789

cubuspl42 commented:
Let's add "N/A" in the Android / iOS "Screenshots/Videos" sections and add a short description of our findings in the "Details" section. I'll proceed with testing in this manner.

cubuspl42 commented:
This is an important line:

https://github.com/Expensify/App/blob/820fb3254eb3fa29152ed4b0df72b85b9c8581a1/src/libs/Browser/index.website.ts#L77

It's a desktop-only logic, which might suggest we're observing the expected behavior

AmjedNazzal commented:
> This is an important line:
> 
> https://github.com/Expensify/App/blob/820fb3254eb3fa29152ed4b0df72b85b9c8581a1/src/libs/Browser/index.website.ts#L77
> 
> It's a desktop-only logic, which might suggest we're observing the expected behavior

You are correct, I've looked at `index.website.ts` and the surrounding components and it does appear like in `beginDeepLinkRedirect` in `App.ts`, it only calls for opening the URL in desktop app and nothing for Android/iOS native app. I will explain all this in the "Details" section like suggested.

AmjedNazzal commented:
@cubuspl42 I have removed the test steps for Android/iOS and removed the video for them as well, also I've added a brief explanation in the Details section.

cubuspl42 commented:
> @cubuspl42 I have removed the test steps for Android/iOS and removed the video for them as well, also I've added a brief explanation in the Details section.

Thank you very much! I'll get back to this tomorrow. I had some high-priority things in my queue today, and it's already evening here.

AmjedNazzal commented:
@cubuspl42 When you have time to look at this again, since you have a physical iphone from what I remember, can you please test opening the URL from iCloud notes? I've seen this issue post about iCloud notes and opening URLs from there into the iOS app and it made me a little bit worried since I've stated there is no natural workflow for deep linking into the iOS app but maybe there is? Thank you!

cubuspl42 commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/2590174/1f03ba8f-04e5-444b-8584-3f3bd04c9bd7

</details>

<details>
<summary>Mobile Web - Chrome</summary>

N/A

</details>

<details>
<summary>Mobile Web - Safari</summary>

N/A

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/2590174/280edceb-877a-4bc9-b1fd-06fe99a6e9b6

</details>

<details>
<summary>iOS</summary>

N/A

</details>

<details>
<summary>Android</summary>

N/A

</details>

cubuspl42 commented:
@francoisl 

The flow we're modifying isn't expected to be handled on mobile native variants of Expensify (iOS / Android). There are no defined deep links for this functionality.

AmjedNazzal committed:
Merge branch 'Expensify:main' into Issue31789

francoisl commented on src/libs/actions/Session/index.ts:
```diff
@@ -332,28 +332,38 @@ function signInWithShortLivedAuthToken(email: string, authToken: string) {
     ];
 
-    const failureData: OnyxUpdate[] = [
+    // Subsequently, we revert it back to the default value of 'signedInWithShortLivedAuthToken' (undefined) in 'successData' or 'failureData' to ensure the user is logged out on refresh
```
Why are we setting `signedInWithShortLivedAuthToken` to `undefined` specifically, rather than `null` so it gets removed?

AmjedNazzal replied:
I actually don't have a specific strong reason for this, we can use null if you feel it's better to remove it rather than keeping it as undefined.

cubuspl42 replied:
I think we should just switch this to `null`. I missed this bit.

AmjedNazzal replied:
Done

AmjedNazzal committed:
Fix issue #31789 changed undefined to null

AmjedNazzal commented:
@francoisl @cubuspl42 The code now resets with `null` instead of `undefined`, I've tested and it seem to be behaving like expected.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7256301720) to staging by https://github.com/francoisl in version: 1.4.14-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7256301720) to staging by https://github.com/francoisl in version: 1.4.14-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7284210064) to production by https://github.com/jasperhuangg in version: 1.4.14-6 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅
